ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt

```diff
@@ -55,88 +58,90 @@ public object AuthenticationChecked : Hook<suspend (ApplicationCall) -> Unit> {
-            if (authenticationContext.principal != null) return@on
-
-            val configurations = pluginConfig.configurations.map { configurationName ->
-                authConfig.providers[configurationName] ?: throw IllegalArgumentException(
-                    if (configurationName == null) {
-                        "Default authentication configuration was not found"
-                    } else {
-                        "Authentication configuration with the name $configurationName was not found"
-                    }
-                )
+public val AuthenticationInterceptors: RouteScopedPlugin<RouteAuthenticationConfig> = createRouteScopedPlugin(
+    "AuthenticationInterceptors",
+    ::RouteAuthenticationConfig
+) {
+    on(AuthenticationHook) { call ->
+        val authConfig = call.application.plugin(Authentication).config
+
+        val authenticationContext = AuthenticationContext.from(call)
+        if (authenticationContext.principal != null) return@on
+
+        val configurations = pluginConfig.configurations.map { configurationName ->
+            authConfig.providers[configurationName] ?: throw IllegalArgumentException(
+                if (configurationName == null) {
+                    "Default authentication configuration was not found"
+                } else {
+                    "Authentication configuration with the name $configurationName was not found"
+                }
+            )
+        }
+        for (provider in configurations) {
+            if (provider.skipWhen.any { skipCondition -> skipCondition(call) }) continue
+
+            provider.onAuthenticate(authenticationContext)
+
+            if (authenticationContext.principal != null) {
+                LOGGER.trace("Authentication succeeded for ${call.request.uri} with provider $provider")
+                break
             }
-            for (provider in configurations) {
-                if (provider.skipWhen.any { skipCondition -> skipCondition(call) }) continue
+        }
+    }
 
-                provider.onAuthenticate(authenticationContext)
+    on(ChallengeHook) { call ->
+        val context = AuthenticationContext.from(call)
 
-                if (authenticationContext.principal != null) break
+        if (context.principal != null) return@on
+        if (context.challenge.completed) {
+            if (!call.isHandled) {
+                call.respond(UnauthorizedResponse())
             }
+            return@on
+        }
+        if (pluginConfig.optional && context.allFailures.none { it == AuthenticationFailedCause.InvalidCredentials }) {
+            return@on
         }
 
-        on(ChallengeHook) { call ->
-            val context = AuthenticationContext.from(call)
+        val challenges = context.challenge.challenges
 
-            if (context.principal != null) return@on
+        for (challenge in challenges) {
+            challenge(context.challenge, call)
```

Comment:
should we add logs inside providers and challenges?

Reply:
let's make it as a separate task